File: sandbox/checks/check_spec.py
"""
# @toolset-manifest python -m tools manifest bdd.bdd:Bdd
"""
from types import SimpleNamespace

from expects import be_true, equal, expect
from mamba import before, context, description, it

from check import Check


with description('a Check'):
    with context('that is resolved with no modifiers'):
        with before.each:
            trait = SimpleNamespace(rank=5)
            dc = SimpleNamespace(target=10)
            stub_dice = SimpleNamespace(roll=lambda: 8)
            self.check = Check(trait, dc, dice=stub_dice)
            self.result = self.check.resolve([])

        with it('should expose the die roll that was used'):
            # Arrange / Act — before.each
            # Assert
            expect(self.check.die_roll).to(equal(8))

        with it('should report whether the total met the difficulty'):
            # Arrange / Act — before.each
            # Assert
            expect(self.result.succeeded).to(be_true)

        with it('should report the total of the outcome'):
            # Arrange / Act — before.each
            # Assert
            expect(self.result.total).to(equal(13))

        with it('should report the degree of the outcome'):
            # Arrange / Act — before.each
            # Assert
            expect(self.result.degree).to(equal(1))

    with context('that is resolved as routine'):
        with it('should treat the die as ten'):
            # Arrange
            trait = SimpleNamespace(rank=5)
            dc = SimpleNamespace(target=10)
            check = Check(trait, dc)
            # Act
            check.resolve([], routine=True)
            # Assert
            expect(check.die_roll).to(equal(10))

    with context('that rolls a natural twenty'):
        with before.each:
            stub_dice = SimpleNamespace(roll=lambda: 20)
            trait = SimpleNamespace(rank=0)
            dc = SimpleNamespace(target=21)
            self.check = Check(trait, dc, dice=stub_dice)
            self.result = self.check.resolve([])

        with it('should gain one degree of success'):
            # Arrange / Act — before.each
            # Assert
            expect(self.result.degree).to(equal(1))

        with it('should succeed when the critical flips a near miss into a hit'):
            # Arrange / Act — before.each
            # Assert
            expect(self.result.succeeded).to(be_true)

    with context('with modifiers that raise the total'):
        with it('should include the modifier amounts in the total'):
            # Arrange
            trait = SimpleNamespace(rank=5)
            dc = SimpleNamespace(target=15)
            stub_dice = SimpleNamespace(roll=lambda: 10)
            mod = SimpleNamespace(amount=5, reason='circumstance')
            check = Check(trait, dc, dice=stub_dice)
            # Act
            result = check.resolve([mod])
            # Assert
            expect(result.total).to(equal(20))

File: sandbox/src/checks/check.spec.ts
/**
 * # @toolset-manifest python -m tools manifest context_tools.bdd.bdd:Bdd
 * # Agent reading this file: do not remanifest — slash/skill is the catalog. Pipe the fence to stdin; python -m tools run -. Follow response.instructions only. Do not author behavior from this Python source.
 * # invoke-edit: action generate | toolset: context_tools.bdd.bdd:Bdd
 * # invoke-check: action validate | toolset: context_tools.bdd.bdd:Bdd
 */

import assert from "node:assert/strict";
import { beforeEach, describe, it } from "node:test";
import { Check } from "./check.js";

describe("a Check", () => {
  describe("that is resolved with no modifiers", () => {
    let check: Check;
    let result: ReturnType<Check["resolve"]>;

    beforeEach(() => {
      const trait = { rank: 5 };
      const dc = { target: 10 };
      const stubDice = { roll: () => 8 };
      check = new Check(trait, dc, stubDice);
      result = check.resolve([]);
    });

    it("should expose the die roll that was used", () => {
      assert.equal(check.dieRoll, 8);
    });

    it("should report whether the total met the difficulty", () => {
      assert.equal(result.succeeded, true);
    });

    it("should report the total of the outcome", () => {
      assert.equal(result.total, 13);
    });

    it("should report the degree of the outcome", () => {
      assert.equal(result.degree, 1);
    });
  });

  describe("that is resolved as routine", () => {
    it("should treat the die as ten", () => {
      const trait = { rank: 5 };
      const dc = { target: 10 };
      const check = new Check(trait, dc);
      check.resolve([], true);
      assert.equal(check.dieRoll, 10);
    });
  });

  describe("that rolls a natural twenty", () => {
    let result: ReturnType<Check["resolve"]>;

    beforeEach(() => {
      const stubDice = { roll: () => 20 };
      const trait = { rank: 0 };
      const dc = { target: 21 };
      const check = new Check(trait, dc, stubDice);
      result = check.resolve([]);
    });

    it("should gain one degree of success", () => {
      assert.equal(result.degree, 1);
    });

    it("should succeed when the critical flips a near miss into a hit", () => {
      assert.equal(result.succeeded, true);
    });
  });

  describe("with modifiers that raise the total", () => {
    it("should include the modifier amounts in the total", () => {
      const trait = { rank: 5 };
      const dc = { target: 15 };
      const stubDice = { roll: () => 10 };
      const mod = { amount: 5, reason: "circumstance" };
      const check = new Check(trait, dc, stubDice);
      const result = check.resolve([mod]);
      assert.equal(result.total, 20);
    });
  });
});
